### PR TITLE
[GTK][WPE] ThreadedCompositor: Switch didRenderFrameTimer to use a RunLoopObserver

### PR DESCRIPTION
**Pull request by nikolaszimmermann**
#### d0fa6302fb1073f1a843a0d3c5bc887d8cbfc37f
<pre>
Unreviewed, reverting <a href="https://commits.webkit.org/302433@main">302433@main</a> (28fa568972a4)
</pre>

### CODE REVIEW

File: Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp [hunk 1]
```diff
@@ -321,12 +320,15 @@ void ThreadedCompositor::renderLayerTree()
...
-    if (!m_didRenderFrameTimer.isActive())
-        m_didRenderFrameTimer.startOneShot(0_s);
 #if !HAVE(OS_SIGNPOST) && !USE(SYSPROF_CAPTURE)
     UNUSED_VARIABLE(compositionRequestID);
 #endif
 
+    ASSERT(!m_didCompositeRunLoopObserver->isScheduled());
+    RunLoop::mainSingleton().dispatch([this, protectedThis = Ref { *this }] {
+        m_didCompositeRunLoopObserver->schedule();
+    });
```
Comment (carlosgcampos) [inline Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp]: schedule is not thread safe? if we have to dispatch a source in the main thread, then the timer was better since it has higher priority. The idea of the observer here is precisely avoid scheduling a source in the main thread run loop.

Comment (carlosgcampos) [reply Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp]: I think we can update RunLoopObserver::schedule to handle the platform run loop. Then, we can schedule from the compositor thread but passing the main run loop.

Comment (nikolaszimmermann) [reply Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp]: When directly scheduling from `renderLayerTree`, we end up calling `flushLayers` from the composition thread, not the main thread. That's why I used the dispatch on main loop hack.  Otherwise:

```
ASSERTION FAILED: canSafelyBeUsed()
WTF/Headers/wtf/WeakRef.h(111) : T *WTF::WeakRef<WebCore::LocalFrame>::operator->() const [T = WebCore::LocalFrame, WeakPtrImpl = WTF::DefaultWeakPtrImpl]
ASSERTION FAILED: Unsafe to ref/deref from different threads
m_isOwnedByMainThread == isMainThread()
WTF/Headers/wtf/RefCounted.h(119) : void WTF::RefCountedBase::applyRefDerefThreadingCheck() const
1   0x79a8987a375a WebCore::Page::layoutIfNeeded(WTF::OptionSet<WebCore::LayoutOptions, (WTF::ConcurrencyTag)0>)
2   0x79a8987a414b WebCore::Page::updateRendering()
3   0x79a893c53390 WebKit::WebPage::updateRendering()
4   0x79a893ca134f WebKit::LayerTreeHost::flushLayers()
5   0x79a893ca2b3f WebKit::LayerTreeHost::didComposite(unsigned int)
6   0x79a893ca8bbe WTF::Detail::CallableWrapper<WebKit::ThreadedCompositor::ThreadedCompositor(WebKit::LayerTreeHost&)::$_1, void>::call()
7   0x79a8989975d2 WTF::Detail::CallableWrapper<WebCore::RunLoopObserver::schedule(void*, WTF::OptionSet<WTF::RunLoop::Activity, (WTF::ConcurrencyTag)0>)::$_0, WTF::ActivityObserver::ContinueObservation>::call()
8   0x79a895bb6b49 WTF::RunLoop::notifyActivity(WTF::RunLoop::Activity)
9   0x79a895bb689d WTF::RunLoop::runGLibMainLoopIteration(WTF::RunLoop::MayBlock)
10  0x79a895bb7095 WTF::RunLoop::run()
11  0x79a895b37408 WTF::Thread::entryPoint(WTF::Thread::NewThreadContext*)
12  0x79a895bbe099 WTF::wtfThreadEntryPoint(void*)
13  0x79a88d759aa4 start_thread
14  0x79a88d7e6a64 __clone
```

Respecting the PlatformRunLoop is nice idea, will explore it.

Comment (nikolaszimmermann) [reply Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp]: Needed to adjust things a bit for cross-thread scheduling, please have a look :-)

Comment (carlosgcampos) [reply Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp]: > When directly scheduling from `renderLayerTree`, we end up calling `flushLayers` from the composition thread, not the main thread. That's why I used the dispatch on main loop hack. Otherwise:
> 
> ```
> ASSERTION FAILED: canSafelyBeUsed()
> WTF/Headers/wtf/WeakRef.h(111) : T *WTF::WeakRef<WebCore::LocalFrame>::operator->() const [T = WebCore::LocalFrame, WeakPtrImpl = WTF::DefaultWeakPtrImpl]
> ASSERTION FAILED: Unsafe to ref/deref from different threads
> m_isOwnedByMainThread == isMainThread()
> WTF/Headers/wtf/RefCounted.h(119) : void WTF::RefCountedBase::applyRefDerefThreadingCheck() const
> 1   0x79a8987a375a WebCore::Page::layoutIfNeeded(WTF::OptionSet<WebCore::LayoutOptions, (WTF::ConcurrencyTag)0>)
> 2   0x79a8987a414b WebCore::Page::updateRendering()
> 3   0x79a893c53390 WebKit::WebPage::updateRendering()
> 4   0x79a893ca134f WebKit::LayerTreeHost::flushLayers()
> 5   0x79a893ca2b3f WebKit::LayerTreeHost::didComposite(unsigned int)
> 6   0x79a893ca8bbe WTF::Detail::CallableWrapper<WebKit::ThreadedCompositor::ThreadedCompositor(WebKit::LayerTreeHost&)::$_1, void>::call()
> 7   0x79a8989975d2 WTF::Detail::CallableWrapper<WebCore::RunLoopObserver::schedule(void*, WTF::OptionSet<WTF::RunLoop::Activity, (WTF::ConcurrencyTag)0>)::$_0, WTF::ActivityObserver::ContinueObservation>::call()
> 8   0x79a895bb6b49 WTF::RunLoop::notifyActivity(WTF::RunLoop::Activity)
> 9   0x79a895bb689d WTF::RunLoop::runGLibMainLoopIteration(WTF::RunLoop::MayBlock)
> 10  0x79a895bb7095 WTF::RunLoop::run()
> 11  0x79a895b37408 WTF::Thread::entryPoint(WTF::Thread::NewThreadContext*)
> 12  0x79a895bbe099 WTF::wtfThreadEntryPoint(void*)
> 13  0x79a88d759aa4 start_thread
> 14  0x79a88d7e6a64 __clone
> ```

Of course, didComposite needs to be called in the main thread, what I suggested was to schedule from the compositor thread an observer on the main thread run loop.

> Respecting the PlatformRunLoop is nice idea, will explore it.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp [hunk 2]
```diff
@@ -29,24 +29,33 @@
 #if USE(GLIB_EVENT_LOOP)
 namespace WebCore {
 
-void RunLoopObserver::schedule(PlatformRunLoop, OptionSet<Activity> activities)
+void RunLoopObserver::schedule(PlatformRunLoop runLoop, OptionSet<Activity> activities)
 {
-    if (m_runLoopObserver)
+    if (!runLoop)
+        runLoop = RunLoop::currentSingleton();
+    ASSERT(runLoop);
```
Comment (carlosgcampos) [inline Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: If the given parameter is nullptr we use currentSingleton that returns a reference, so I don't think this assert is needed and it's confusing.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp [hunk 3]
```diff
@@ -29,24 +29,33 @@
...
+    }
 
     m_runLoopObserver = ActivityObserver::create(static_cast<uint8_t>(m_order), activities, [this]() {
         runLoopObserverFired();
         return isRepeating() ? ActivityObserver::ContinueObservation::Yes : ActivityObserver::ContinueObservation::No;
     });
 
-    RunLoop::currentSingleton().observeActivity(*m_runLoopObserver);
+    m_scheduledOnRunLoop = runLoop;
+    m_scheduledOnRunLoop->observeActivity(*m_runLoopObserver);
```
Comment (carlosgcampos) [inline Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: Since we always observe right after create and unobserve right before destroy, why don't we handle the run loop from ActivityObserver? We pass it to the constructor, so that the ActivityObserver is always attached to a run loop and it internally calls observe from constructor and unobserve from destructor.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp (hunk 2 above)
Comment (nikolaszimmermann) [reply Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: Heh, indeed. I had the assert there before adding the if (!runLoop), until the assertion fired, as I forgot about the default value being nullptr :-)

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp (hunk 3 above)
Comment (nikolaszimmermann) [reply Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: Will explore that.

File: Source/WTF/wtf/glib/ActivityObserver.h [hunk 4]
```diff
@@ -21,37 +21,52 @@
...
 
+    ~ActivityObserver()
+    {
+        m_runLoop->unobserveActivity(*this);
+    }
+
+    void attachObserverToRunLoop()
+    {
+        m_runLoop->observeActivity(*this);
+    }
```
Comment (carlosgcampos) [inline Source/WTF/wtf/glib/ActivityObserver.h]: We could make this private and call this from create() since it's the only way to create an ActivityObserver it will always be called. This way the caller doesn't need to care about attaching.

Comment (nikolaszimmermann) [reply Source/WTF/wtf/glib/ActivityObserver.h]: Neat.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp [hunk 5]
```diff
@@ -29,25 +29,27 @@
...
+    if (!runLoop)
+        runLoop = RunLoop::currentSingleton();
+
+    if (m_runLoopObserver) {
+        ASSERT(&m_runLoopObserver->runLoop() == runLoop.get());
         return;
+    }
 
-    m_runLoopObserver = ActivityObserver::create(static_cast<uint8_t>(m_order), activities, [this]() {
+    m_runLoopObserver = ActivityObserver::create(Ref { *runLoop.get() }, static_cast<uint8_t>(m_order), activities, [this]() {
```
Comment (carlosgcampos) [inline Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: Can we use releaseNonNull()?

File: Source/WTF/wtf/glib/ActivityObserver.h [hunk 6]
```diff
@@ -21,37 +21,54 @@
...
+    RunLoop& runLoop() const { return m_runLoop; }
     uint8_t order() const { return m_order; }
     OptionSet<RunLoop::Activity> activities() const { return m_activities; }
 
     ContinueObservation notify() const { return m_callback(); }
 
 private:
-    ActivityObserver(uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
-        : m_order(order)
+    friend Ref<ActivityObserver> create(Ref<RunLoop>&&, uint8_t, OptionSet<RunLoop::Activity>, Callback&&);
```
Comment (carlosgcampos) [inline Source/WTF/wtf/glib/ActivityObserver.h]: Why do we need this?

**Review by carlosgcampos — APPROVED** [review]

File: Source/WTF/wtf/glib/ActivityObserver.h (hunk 6 above)
Comment (nikolaszimmermann) [reply Source/WTF/wtf/glib/ActivityObserver.h]: Oops leftover, from when I called attachObserver.. from create(). Will remove.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp [hunk 7]
```diff
@@ -29,23 +29,28 @@
...
 
-    RunLoop::currentSingleton().observeActivity(*m_runLoopObserver);
+    m_runLoopObserver->addObserverToRunLoop();
 }
 
 void RunLoopObserver::invalidate()
 {
     if (m_runLoopObserver) {
-        RunLoop::currentSingleton().unobserveActivity(*m_runLoopObserver);
+        m_runLoopObserver->removeObserverFromRunLoop();
```
Comment (carlosgcampos) [inline Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: I think we need to protect m_runLoopObserver with a lock, because we call schedule and invalidate from different threads

File: Source/WTF/wtf/glib/ActivityObserver.h [hunk 8]
```diff
@@ -26,32 +26,45 @@
...
+        : m_runLoop(WTFMove(runLoop))
+        , m_order(order)
         , m_activities(activities)
         , m_callback(WTFMove(callback))
     {
         ASSERT(m_callback);
     }
 
 private:
+    Ref<RunLoop> m_runLoop;
```
Comment (carlosgcampos) [inline Source/WTF/wtf/glib/ActivityObserver.h]: Could you remind me why we don't use a thread safe weak ptr? The run loop keeps a reference to the activity observer and the activity observer to the run loop, isn't this a circular dependency? We don't need to keep the run loop alive only to unobserve the activity, no?

File: Source/WTF/wtf/glib/ActivityObserver.h [hunk 9]
```diff
@@ -26,32 +26,45 @@
...
     using Callback = Function<ContinueObservation()>;
 
-    static Ref<ActivityObserver> create(uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
+    static Ref<ActivityObserver> create(Ref<RunLoop>&& runLoop, uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
     {
-        return adoptRef(*new ActivityObserver(order, activities, WTFMove(callback)));
+        return adoptRef(*new ActivityObserver(WTFMove(runLoop), order, activities, WTFMove(callback)));
     }
 
+    void addObserverToRunLoop()
```
Comment (carlosgcampos) [inline Source/WTF/wtf/glib/ActivityObserver.h]: From a method called addObserverToRunLoop I would expect a run loop to be passed as parameter, but since the observer already knows what run loop to attach, what about using just attach/detach, or start/stop?

Comment (nikolaszimmermann) [reply Source/WTF/wtf/glib/ActivityObserver.h]: Good idea, attach/detach.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp (hunk 7 above)
Comment (nikolaszimmermann) [reply Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: Good catch, yikes.

File: Source/WTF/wtf/glib/ActivityObserver.h (hunk 8 above)
Comment (nikolaszimmermann) [reply Source/WTF/wtf/glib/ActivityObserver.h]: Done.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp [hunk 10]
```diff
@@ -29,29 +29,33 @@
...
 
-    m_runLoopObserver = ActivityObserver::create(static_cast<uint8_t>(m_order), activities, [this]() {
+    auto effectiveRunLoop = runLoop.get() ? runLoop : ThreadSafeWeakPtr { RunLoop::currentSingleton() };
+    m_runLoopObserver = ActivityObserver::create(WTFMove(effectiveRunLoop), static_cast<uint8_t>(m_order), activities, [this]() {
         runLoopObserverFired();
         return isRepeating() ? ActivityObserver::ContinueObservation::Yes : ActivityObserver::ContinueObservation::No;
     });
 
-    RunLoop::currentSingleton().observeActivity(*m_runLoopObserver);
+    m_runLoopObserver->attachObserverToRunLoop();
```
Comment (carlosgcampos) [inline Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: I meant to remove the ToRunLoop() since that sounds like we are passing the run loop to attach too, but ActivityObserver already knows about the run loop. And also Observer, attachObserver sounds like attaching another thing, but the class is already the observer, so I think m_runLoopObserver->attach() or maybe better m_runLoopObserver->start() because at that point we start observing.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp [hunk 11]
```diff
@@ -29,29 +29,33 @@
...
     if (m_runLoopObserver) {
-        RunLoop::currentSingleton().unobserveActivity(*m_runLoopObserver);
+        m_runLoopObserver->detachObserverFromRunLoop();
         m_runLoopObserver = nullptr;
     }
 }
 
 bool RunLoopObserver::isScheduled() const
 {
+    Locker lock { const_cast<RunLoopObserver&>(*this).m_runLoopObserverLock };
```
Comment (carlosgcampos) [inline Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: We normally make the lock mutable to be able to use it in const methods

File: Source/WebCore/platform/RunLoopObserver.h [hunk 12]
```diff
@@ -33,12 +33,17 @@
...
+#include <wtf/WeakPtr.h>
 #include <wtf/glib/ActivityObserver.h>
 #endif
 
 #if USE(CF)
 using PlatformRunLoopObserver = struct __CFRunLoopObserver*;
 using PlatformRunLoop = struct __CFRunLoop*;
+#elif USE(GLIB)
+using PlatformRunLoopObserver = ActivityObserver;
+using PlatformRunLoop = ThreadSafeWeakPtr<RunLoop>;
```
Comment (carlosgcampos) [inline Source/WebCore/platform/RunLoopObserver.h]: I wouldn't expose the WeakPtr detail at this point, it can be a RefPtr and then ActivityObserver creates the WeakPtr at construction.

Comment (nikolaszimmermann) [reply Source/WebCore/platform/RunLoopObserver.h]: Ok.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp (hunk 11 above)
Comment (nikolaszimmermann) [reply Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: That's better indeed.

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp (hunk 10 above)
Comment (nikolaszimmermann) [reply Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: Let's go for start/stop - I understand and agree with the reasoning.

File: Source/WebCore/platform/RunLoopObserver.h [hunk 13]
```diff
@@ -33,12 +33,17 @@
 #include <wtf/TZoneMalloc.h>
 
 #if USE(GLIB_EVENT_LOOP)
+#include <wtf/Lock.h>
+#include <wtf/WeakPtr.h>
```
Comment (carlosgcampos) [inline Source/WebCore/platform/RunLoopObserver.h]: We don't need this here

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp [hunk 14]
```diff
@@ -29,29 +29,33 @@
...
-void RunLoopObserver::schedule(PlatformRunLoop, OptionSet<Activity> activities)
+void RunLoopObserver::schedule(PlatformRunLoop runLoop, OptionSet<Activity> activities)
 {
+    Locker lock { m_runLoopObserverLock };
     if (m_runLoopObserver)
         return;
 
-    m_runLoopObserver = ActivityObserver::create(static_cast<uint8_t>(m_order), activities, [this]() {
+    ThreadSafeWeakPtr weakRunLoop { runLoop ? *runLoop.get() : RunLoop::currentSingleton() };
+    m_runLoopObserver = ActivityObserver::create(WTFMove(weakRunLoop), static_cast<uint8_t>(m_order), activities, [this]() {
```
Comment (carlosgcampos) [inline Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: I meant ActivityObserver receives a ref, because we never pass nullptr, and the constructor creates the WeakPtr, because it's an implementation detail of ActivityObserver.

File: Source/WTF/wtf/glib/ActivityObserver.h [hunk 15]
```diff
@@ -43,15 +56,17 @@ class ActivityObserver : public RefCounted<ActivityObserver> {
     ContinueObservation notify() const { return m_callback(); }
 
 private:
-    ActivityObserver(uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
-        : m_order(order)
+    ActivityObserver(ThreadSafeWeakPtr<RunLoop>&& runLoop, uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
+        : m_runLoop(WTFMove(runLoop))
```
Comment (carlosgcampos) [inline Source/WTF/wtf/glib/ActivityObserver.h:78]: Here is where we should create the WeakPtr

**Review by carlosgcampos — APPROVED** [review]

File: Source/WebCore/platform/glib/RunLoopObserverGLib.cpp (hunk 14 above)
Comment (nikolaszimmermann) [reply Source/WebCore/platform/glib/RunLoopObserverGLib.cpp]: Ok, I didn't get that.

**Comment by nikolaszimmermann** [comment]
After all the difficulties with the threading corner cases, add some API tests related to threading + RunLoopObserver. I'm open for suggestions how to improve them.

**Review by carlosgcampos — APPROVED** [review]
